runtime/kphp_tracing_binlog.cpp

```diff
@@ -0,0 +1,258 @@
+// that new strings are flushed separately;
+// there are 2 ways to flush them:
+// 1) after every request that had registered new strings, push them into a file
+// 2) keep them in memory and push all at once just before any trace that will be flushed later
+// the case (2) is what we are focusing on; we are keeping new strings in memory until any other trace flushed -
+// but not as "string" (that would require copying from script memory), but as a separate full-fledged binlog
+// with events type=etStringsTableRegister and inlined strings;
+// in kphp_tracing.cpp, there is a variable:
+// `static tracing_binary_buffer flush_strings_binlog;`
+// if a trace is not flushed, `flush_strings_binlog.append_current_php_script_strings()` is called;
+// when inserting to that binlog, we take memory from heap using dl::heap_allocate()
+// so that it outlives current script execution;
+// before any trace flush, if that binlog is not empty, it's flushed in advance and cleared
+static std::vector<string> current_php_script_registered_strings;
+
+
+void tracing_binary_buffer::set_use_heap_memory() {
+  use_heap_memory = true;
+}
+
+void tracing_binary_buffer::init_and_alloc() {
+  finish_cur_chunk_start_next();
+}
+
+void tracing_binary_buffer::alloc_next_chunk_if_not_enough(int reserve_bytes) {
+  if (unlikely(last_chunk == nullptr)) {
+    php_critical_error("tracing not initialized, but used");
+  }
+  int cur_size = get_cur_chunk_size();
+  if (cur_size + reserve_bytes > BUF_CHUNK_SIZE) {
+    finish_cur_chunk_start_next();
+  }
+}
+
+void tracing_binary_buffer::finish_cur_chunk_start_next() {
+  if (last_chunk != nullptr) {
+    last_chunk->size_bytes = get_cur_chunk_size();
+  }
+
+  char *mem = reinterpret_cast<char *>(use_heap_memory ? dl::heap_allocate(BUF_CHUNK_SIZE) : dl::allocate(BUF_CHUNK_SIZE));
+  one_chunk *chunk = reinterpret_cast<one_chunk *>(mem);
+  chunk->buf = reinterpret_cast<int *>(mem + sizeof(one_chunk));
+  chunk->size_bytes = 0;
+  chunk->prev_chunk = last_chunk;
+
+  last_chunk = chunk;
+  pos = last_chunk->buf;
+}
+
+void tracing_binary_buffer::clear(bool real_deallocate) {
+  // for script memory, no need of dl::deallocate(), because clear() is called at the end of the script,
+  // we just need to reset pointers, so that dl::allocate() is called again for a new script;
+  // but for heap memory, we need to really deallocate all chunks
+
+  if (real_deallocate) {
+    for (one_chunk *chunk = last_chunk; chunk;) {
+      one_chunk *prev_chunk = chunk->prev_chunk;
+      dl::heap_deallocate(chunk, BUF_CHUNK_SIZE);
+      chunk = prev_chunk;
+    }
+  }
+
+  last_chunk = nullptr;
+  pos = nullptr;
+}
+
+void tracing_binary_buffer::write_event_type(EventTypeEnum event_type, int custom24bits) {
+  if (unlikely(custom24bits < 0 || custom24bits >= 1 << 24)) {
+    php_warning("custom24bits %d overflow next to event_type %d", custom24bits, static_cast<int>(event_type));
+  }
+  alloc_next_chunk_if_not_enough(128);
```

Comment:
Pretty sure. `write_event_type()` is called for every event, (see `BinlogWriter` methods), all events are tiny in fact, 24 bytes max. If an event will contain an inlined string (not an indexed one), `write_string_inlined()` will also check size.

Checks inside `write_int()` and other primitive functions aren't presented intentionally, because `write_event_type()` allocates enough space. 

I'll add a comment above this line.